app/permissions_migrations.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2018-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import "github.com/mattermost/mattermost-server/model"
+
+type permissionTransformation struct {
+	On     func(map[string]bool) bool
+	Add    []string
+	Remove []string
+}
+type permissionsMap []permissionTransformation
+
+const (
+	MIGRATION_KEY_EMOJI_PERMISSIONS_SPLIT   = "emoji_permissions_split"
+	MIGRATION_KEY_WEBHOOK_PERMISSIONS_SPLIT = "webhook_permissions_split"
+
+	PERMISSION_MANAGE_EMOJIS                   = "manage_emojis"
+	PERMISSION_MANAGE_OTHERS_EMOJIS            = "manage_others_emojis"
+	PERMISSION_CREATE_EMOJIS                   = "create_emojis"
+	PERMISSION_DELETE_EMOJIS                   = "delete_emojis"
+	PERMISSION_DELETE_OTHERS_EMOJIS            = "delete_others_emojis"
+	PERMISSION_MANAGE_WEBHOOKS                 = "manage_webhooks"
+	PERMISSION_MANAGE_OTHERS_WEBHOOKS          = "manage_others_webhooks"
+	PERMISSION_MANAGE_INCOMING_WEBHOOKS        = "manage_incoming_webhooks"
+	PERMISSION_MANAGE_OTHERS_INCOMING_WEBHOOKS = "manage_others_incoming_webhooks"
+	PERMISSION_MANAGE_OUTGOING_WEBHOOKS        = "manage_outgoing_webhooks"
+	PERMISSION_MANAGE_OTHERS_OUTGOING_WEBHOOKS = "manage_others_outgoing_webhooks"
+)
+
+func permissionExists(permission string) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		val, ok := permissions[permission]
+		return ok && val
+	}
+}
+
+func permissionNotExists(permission string) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		val, ok := permissions[permission]
+		return !(ok && val)
+	}
+}
+
+func permissionOr(funcs ...func(map[string]bool) bool) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		for _, f := range funcs {
+			if f(permissions) {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+func permissionAnd(funcs ...func(map[string]bool) bool) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		for _, f := range funcs {
+			if !f(permissions) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+func applyPermissionsMap(permissions []string, migrationMap permissionsMap) []string {
+	finalMap := make(map[string]bool)
+	var result []string
+	for _, permission := range permissions {
+		finalMap[permission] = true
+	}
+
+	for _, transformation := range migrationMap {
+		if transformation.On(finalMap) {
+			for _, add := range transformation.Add {
+				finalMap[add] = true
+			}
+			for _, remove := range transformation.Remove {
+				finalMap[remove] = false
+			}
+		}
+	}
+
+	for key, active := range finalMap {
+		if active {
+			result = append(result, key)
+		}
+	}
+	return result
+}
```

Comment:
Took a bit of thinking initially to understand, but I actually really really like the functional way you've done this. It's really easy to reason about once I had satisfied myself I trusted the function implementations :)